### ISSUE TITLE
[HOLD for payment 2022-10-20] [$250] Console error - Non-unique keys when group users have same name - reported by @rushatgabhane

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#10847]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login on web dev
2. Create a new group with two users that have the same first name.
3. Send a message

## Expected Result:
No console error

## Actual Result:
Console error - Encountered two children with the same key, `Rushat`. Keys should be

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web (Dev only)

**Version Number:** 1.1.96-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
<img width="1130" alt="image - 2022-09-06T234720 499" src="https://user-images.githubusercontent.com/43995119/188784466-60a29d29-a5f2-4053-be83-b4ebcc37d824.png">

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @rushatgabhane
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1661196448723269

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by laurenreidexpensify** [comment]
@rushatgabhane I'm not sure I can reproduce this - is there something I'm missing from the repro steps? 

<img width="1398" alt="2022-09-07_08-33-31" src="https://user-images.githubusercontent.com/62073721/188817288-a08b6997-a4df-45a8-b945-63976ba6bc02.png">

**Comment by laurenreidexpensify** [comment]
@rushatgabhane nah I'm only on prod and staging. Can you edit the repro steps to make it clear it's only reproducible there? Thanks :)

**Comment by arielgreen** [comment]
https://www.upwork.com/jobs/~01dfdf91764d8f970f

**Comment by ajaynagwani777** [comment]
`                    {_.map(displayNamesWithTooltips, ({
                        displayName, pronouns, tooltip,
                    }, index) => (
                        <Text key={displayName}>`


Replace key={displayName} with key={`${displayName}-${index}`}

**Comment by caphamza** [comment]
### Proposal

React uses the **key** prop to understand the component-to-DOM Element relation, so we don't have to worry about what changes on every render. **key** should always remains unique, else React will most probably mix up the elements and mutate the incorrect one. 

The **key** should remain static across re-renders to maintain the performance. so the best practice is to used _index_ as **key**. This way **key** remains static and unique.

hope it helps :)

**Comment by ajaynagwani777** [comment]
**Reply**

Keys should not be kept as indexes as it might have negative impact on the performance. The best way is to use a string that uniquely identifies the list item. Refer this - https://reactjs.org/docs/lists-and-keys.html

**Comment by ajaynagwani777** [comment]
@rushatgabhane can we start with the above proposal?

**Comment by jeet-dhandha** [comment]
### Proposal
- Current proposal solves issue for same display name on active chat. 
- Also resolves overall same display name Issue. Even if there are `2` - `RoomHeaderAvatars` in same view.

<img width="1393" alt="Screenshot 2022-09-17 at 3 00 38 PM" src="https://user-images.githubusercontent.com/78416198/190850133-49029e8e-bd61-46b6-af4e-f92ffc46b47e.png">

Please check Rushat.
And also long time no see ✋ 😄.

**Comment by rushatgabhane** [comment]
Hey 👋

@mananjadhav will be reviewing your proposals here

**Comment by hoangndm** [comment]
**Proposal**

- The current proposal solves the issue of the same display name on active chat. You should change key={firstName} to key={firstName + (new Date().getTime()).toString()}

**Comment by jeet-dhandha** [comment]
> **Proposal**
> 
> * The current proposal solves the issue of the same display name on active chat. You should change key={firstName} to key={firstName + (new Date().getTime()).toString()}

@hoangndm 
Please check below link why we shouldn't use `new Date().getTime()`.
Link : https://stackoverflow.com/a/54168287/17278253

**Comment by mananjadhav** [comment]
I agree with `index` not the best way to go about it, but it is better than having dynamic values like `Math.rand` and `new Date`.

I am good with @mdneyazahmad's proposal [here](https://github.com/Expensify/App/issues/10847#issuecomment-1238975662) which uses `tooltip`. Tooltip is based on the login and makes it unique good enough.

@mdneyazahmad I also checked the other console.log error that you mentioned, I am of the opinion if we should handle in this PR, but if you feel we should log a separate issue that's fine by me as well. 

@srikarparsi All yours.

🎀 👀 🎀 
C+ reviewed

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by srikarparsi** [comment]
The proposal looks good to me but @mvtglobally did you try reproducing on dev or was it on staging/prod? If it was staging/prod could you try it on dev to see if the issue still persists?

**Comment by srikarparsi** [comment]
bump @mvtglobally

**Comment by mvtglobally** [comment]
@srikarparsi sorry for delay. We checked only staging and prod. Don't have dev access to check

**Comment by mananjadhav** [comment]
@srikarparsi I can confirm this exists and I can reproduce this on dev.

**Comment by srikarparsi** [comment]
sounds good @mananjadhav, the tooltip proposal looks good to me as well, lets move forward with that. The second issue is very similar so I agree that we can handle it in this pr

**Comment by mananjadhav** [comment]
@srikarparsi Can you assign the issue/job so that @mdneyazahmad can work on it.

**Comment by srikarparsi** [comment]
yup, sorry I missed that, assigning him now

**Comment by mananjadhav** [comment]
Afaik, @mdneyazahmad you should get the reporting bonus.

cc - @srikarparsi @arielgreen

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mananjadhav** [comment]
@mdneyazahmad, @mvtglobally wont have the dev setup and that’s fine. You can continue with the PR and I’ll take care of the QA partt

**Comment by arielgreen** [comment]
@rushatgabhane @mananjadhav @mdneyazahmad check your Upwork inboxes, please.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests in Staging. (Third week)

**Comment by mananjadhav** [comment]
Waiting on @srikarparsi to review and merge the PR

**Comment by mananjadhav** [comment]
PR is merged. @srikarparsi @arielgreen considering this won't be reproducible on staging and dev. So we won't need a QA here.  I am assuming we would handle the payment the same way?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fourth week)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fifth week) Are we ok to close this issue?

**Comment by srikarparsi** [comment]
keeping the issue open for payment on the 20th

**Comment by arielgreen** [comment]
Paid.

**Comment by mananjadhav** [comment]
Hey @arielgreen,
Quick thing, I did get paid, and I am assuming @rushatgabhane got paid too.

Did we payout @mdneyazahmad for the code as well as bug reporting? We first attempted to resolve two separate bugs in this PR, and one of which @mdneyazahmad had [reported](https://github.com/Expensify/App/issues/10847#issuecomment-1238977292) and I posted a comment about the payment [here](https://github.com/Expensify/App/issues/10847#issuecomment-1257521851)
